metricbeat/module/kafka/broker.go

```diff
@@ -290,6 +299,86 @@ func (b *Broker) FetchPartitionOffsetFromTheLeader(topic string, partitionID int
 	return offset, nil
 }
 
+// FetchPartitionOffsetsForTopics fetches offsets of all partitions from its leaders.
+func (b *Broker) FetchPartitionOffsetsForTopics(topics []*sarama.TopicMetadata, time int64) map[string]map[int32]*PartitionOffsets {
+	leaderTopicPartition, leaderBrokers := b.groupBrokersPerTopicPartitions(topics)
+	defer b.closeLeaders(leaderBrokers)
+
+	topicPartitionPartitionOffsets := b.fetchGroupedPartitionOffsetsPerBroker(leaderTopicPartition, leaderBrokers, time)
+	return topicPartitionPartitionOffsets
+}
+
+func (b *Broker) groupBrokersPerTopicPartitions(topics []*sarama.TopicMetadata) (brokerTopicPartitionsMap, brokersMap) {
+	leaderTopicPartition := brokerTopicPartitionsMap{}
+	leaderBrokers := brokersMap{}
+
+	for _, topic := range topics {
+		for _, partition := range topic.Partitions {
+			broker, err := b.client.Leader(topic.Name, partition.ID)
+			if err != nil {
+				log.Errorf("connecting to leader broker failed (topicName: %s, partitionID: %d): %v",
+					topic.Name, partition.ID, err)
+				continue
+			}
+
+			if _, ok := leaderTopicPartition[broker.ID()]; !ok {
+				leaderTopicPartition[broker.ID()] = map[string]int32{}
+			}
+			leaderTopicPartition[broker.ID()][topic.Name] = partition.ID
+			leaderBrokers[broker.ID()] = broker
+		}
+	}
+	return leaderTopicPartition, leaderBrokers
+}
+
+func (b *Broker) fetchGroupedPartitionOffsetsPerBroker(leaderTopicPartition brokerTopicPartitionsMap,
+	leaderBrokers brokersMap, time int64) map[string]map[int32]*PartitionOffsets {
+	topicPartitionPartitionOffsets := map[string]map[int32]*PartitionOffsets{}
+
+	for leader, topicPartition := range leaderTopicPartition {
+		req := new(sarama.OffsetRequest)
+		for topic, partition := range leaderTopicPartition[leader] {
+			req.AddBlock(topic, partition, time, 1)
+		}
+
+		resp, err := leaderBrokers[leader].GetAvailableOffsets(req)
```

Comment:
Nit and probably unneeded optimizatiod :smile:: we could parallelize requests per leader.

Reply:
Fixed.